### ISSUE TITLE
[Linux] Widgets are not read by Screen Reader when focused

### ISSUE THREAD

**Issue by crentava** [flutter/flutter#95230]
# issue

there was a similar issue #94791 that discusses this but for a screen reader called NVDA. If you use the tab/arrow keys, you are moved to the UI element that your cursor is positioned over, but the screen reader does not seem to recognize that you moved to the element. This makes applications inaccessible because the user interface is not read to a screen reader user.
## Steps to Reproduce

1. Create the default flutter project by typing "flutter create"
2. Run your project with "flutter run."
3. Enable the Orca screen reader by typing "orca" in your terminal than, go back to the application window.
4. Try hitting tab a couple of times, after that, try the same with each of the arrow keys a couple of times.
**Expected results:** 
if you hit tab once, or if you tap one of the arrow keys, the button should be announced by orca
**Actual results:** 
nothing is read
<details>
<summary>Code sample</summary>

This effects any UI made for the Linux desktop in flutter. You are able to test this by making a new project with "flutter create" like I said above.

</details>

<details>
  <summary>Logs</summary>

```
[✓] Flutter (Channel master, 2.6.0-12.0.pre.1023, on Linux Mint 20.2 5.4.0-91-generic, locale en_US.UTF-8)
    • Flutter version 2.6.0-12.0.pre.1023 at /home/sage/snap/flutter/common/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5691671194 (48 minutes ago), 2021-12-13 20:14:06 -0500
    • Engine revision d8c05122d2
    • Dart version 2.16.0 (build 2.16.0-85.0.dev)
    • DevTools version 2.9.1

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.


[✗] Chrome - develop for the web (Cannot find Chrome executable at google-chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Linux toolchain - develop for Linux desktop
    • clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
    • cmake version 3.10.2
    • ninja version 1.8.2
    • pkg-config version 0.29.1

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).

[✓] VS Code (version 1.62.3)
    • VS Code at /usr/share/code
    • Flutter extension version 3.29.0

[✓] Connected device (1 available)
    • Linux (desktop) • linux • linux-x64 • Linux Mint 20.2 5.4.0-91-generic

! Doctor found issues in 3 categories.
```

</details>

**Comment by maheshmnj** [comment]
Hi @crentava, thanks for filing the issue. I am able to reproduce the issue. With Orca enabled all keys are read (including tab) except  `arrow keys` and `enter key` on the Numpad.  This issue is reproducible on stable and the master channel.

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.8.0, on Ubuntu 20.04.3 LTS 5.11.0-41-generic, locale en_GB.UTF-8)
    • Flutter version 2.8.0 at /home/mahesh/Development/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision cf44000065 (6 days ago), 2021-12-08 14:06:50 -0800
    • Engine revision 40a99c5951
    • Dart version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /home/mahesh/Android/Sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /home/mahesh/Downloads/android-studio-2020.3.1.25-linux/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] Android Studio (version 2020.3)
    • Android Studio at /home/mahesh/Downloads/android-studio-2020.3.1.25-linux/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.62.3)
    • VS Code at /usr/share/code
    • Flutter extension version 3.29.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 20.04.3 LTS 5.11.0-41-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 96.0.4664.93

• No issues found!
```

```
[✓] Flutter (Channel stable, 2.8.0, on Ubuntu 20.04.3 LTS 5.11.0-41-generic, locale en_GB.UTF-8)
    • Flutter version 2.8.0 at /home/mahesh/Development/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision cf44000065 (6 days ago), 2021-12-08 14:06:50 -0800
    • Engine revision 40a99c5951
    • Dart version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /home/mahesh/Android/Sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /home/mahesh/Downloads/android-studio-2020.3.1.25-linux/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] Android Studio (version 2020.3)
    • Android Studio at /home/mahesh/Downloads/android-studio-2020.3.1.25-linux/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.62.3)
    • VS Code at /usr/share/code
    • Flutter extension version 3.29.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 20.04.3 LTS 5.11.0-41-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 96.0.4664.93

• No issues found!
```
</details>

**Comment by crentava** [comment]
I don’t mean that the keys are not read. I mean things like buttons and other UI elements are not read when hitting any of the keys

**Comment by maheshmnj** [comment]
> just in case I wasn’t clear before
I don’t mean that the keys are not read. I mean things like buttons and other UI elements are not read when hitting any of the keys

Thanks for the clarification. I can confirm this I have updated my response.

**Comment by HansMuller** [comment]
CC @cbracken, @chunhtai

**Comment by chunhtai** [comment]
cc @robert-ancell , who wrote most of the linux accessibility code.

**Comment by robert-ancell** [comment]
I'm also reproducing this. It appears that Flutter apps aren't being exported on AT-SPI (accessibility) bus at all. This can be seen by running accerciser and not seeing the Flutter app showing there. I haven't been able to get a recent Flutter engine built to work out what is going wrong, but will look at this in the new year.

**Comment by cbracken** [comment]
I'll take a look at bisecting this to an engine commit.

**Comment by maheshmnj** [comment]
Is there a known flutter version where this worked in the past?
I tried on stable 2.2.3 and it doesn't work there either.

**Comment by robert-ancell** [comment]
Trying again today shows the app showing in accerciser sometimes. Restarting accerciser and disabling/enabling screen reader (orca) seems to fix this. This is in both Flutter stable and master channel. Still not having Orca read out any labels for the buttons (though accerciser shows them correctly).

**Comment by robert-ancell** [comment]
I've tried a simple GTK application that has the same behaviour as the default Flutter app - this is not always showing in accerciser, so that suggests this might not be related to Flutter. I also had the button in the Flutter app being announced by Orca some of the time during testing. So this may be the a11y stack being unreliable, and possibly the Flutter implementation triggering issues in that?

**Comment by cbracken** [comment]
I'm not able to reproduce this on the internal Debian-based distro used at Google. I'm using Orca 41.0, libgtk 3.24.30.

I'm using accersiser 3.38.0.

**Comment by maheshmnj** [comment]
@cbracken I am using `orca 3.36.2` on ubuntu 20.04.3 LTS

> mahesh@mahesh:\~$ gtk-launch --version
3.24.20
mahesh@mahesh:\~$ orca --version
3.36.2
mahesh@mahesh:~$

Let me know if you need any additional info.

**Comment by robert-ancell** [comment]
I'm currently using Ubuntu 22.04:
- libgtk-3-0 3.24.30-3ubuntu2
- at-spi2-core 2.42.0-2ubuntu3
- accerciser 3.38.0-1
- orca 41.0-1ubuntu1

So much the same as you @cbracken.

**Comment by cbracken** [comment]
@robert-ancell any further progress on looking into whether this is a deeper GTK issue or a flutter-specific issue?

**Comment by robert-ancell** [comment]
No progress yet but will continue to look into it.

**Comment by robert-ancell** [comment]
I've proposed some fixes for the Linux a11y code:
https://github.com/flutter/engine/pull/31601
https://github.com/flutter/engine/pull/31582
https://github.com/flutter/engine/pull/31602
https://github.com/flutter/engine/pull/31600

These don't specifically relate to the issue here, but perhaps they are confusing Orca and stopping it from working correctly.

**Comment by cbracken** [comment]
@robert-ancell does https://github.com/flutter/engine/pull/32499 have any effect on this?

**Comment by cbracken** [comment]
@crentava @maheshmnj we believe flutter/engine#32499 may possibly have fixed this; does this still repro on Flutter stable branch?

**Comment by maheshmnj** [comment]
@cbracken I can still reproduce it on stable and the master channel. with orca `3.36.2`. On Running the counter app with orca running in a terminal, Focusing the cursor on text widgets or buttons doesn't read anything and Neither does pressing the tab make a difference. 

 I do get some errors with orca not sure if that is causing the issue. Let me know if you want me to check anything.

### logs
```
mahesh@mahesh:~$ orca -v
3.36.2
mahesh@mahesh:~$ orca

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/orca/generator.py", line 243, in generate
    result = eval(formatting, globalsDict)
  File "<string>", line 1, in <module>
NameError: name 'newAncestors' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/orca/generator.py", line 254, in generate
    globalsDict[arg] = self._methodsDict[arg](obj, **args)
  File "/usr/lib/python3/dist-packages/orca/scripts/toolkits/Chromium/speech_generator.py", line 55, in _generateNewAncestors
    return super()._generateNewAncestors(obj, **args)
  File "/usr/lib/python3/dist-packages/orca/scripts/web/speech_generator.py", line 67, in _generateNewAncestors
    return super()._generateNewAncestors(obj, **args)
  File "/usr/lib/python3/dist-packages/orca/speech_generator.py", line 2110, in _generateNewAncestors
    result = self._generateAncestors(obj, **args)
  File "/usr/lib/python3/dist-packages/orca/scripts/web/speech_generator.py", line 82, in _generateAncestors
    result = [super()._generateName(doc)]
  File "/usr/lib/python3/dist-packages/orca/speech_generator.py", line 138, in _generateName
    role = args.get('role', obj.getRole())
AttributeError: 'NoneType' object has no attribute 'getRole'

18:24:13.313005 - EVENT MANAGER: No focus
^CKilled
```

### flutter doctor 
```
mahesh@mahesh:~/Desktop/counter$ flutter --version
Flutter 3.0.4 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 85684f9300 (11 days ago) • 2022-06-30 13:22:47 -0700
Engine • revision 6ba2af10bb
Tools • Dart 2.17.5 • DevTools 2.12.2
mahesh@mahesh:~/Desktop/counter$ flutterm --version
Flutter 3.1.0-0.0.pre.1594 • channel master • https://github.com/flutter/flutter.git
Framework • revision 082ad11422 (3 hours ago) • 2022-07-11 11:40:07 +0200
Engine • revision 2b5d5c3b2d
Tools • Dart 2.18.0 (build 2.18.0-271.0.dev) • DevTools 2.15.0
```

**Comment by jpnurmi** [comment]
Hi @maheshmnj, Orca doesn't have "Speak object under mouse" enabled by default - is this a system that has been previously used for testing with Orca? What happens if you force Orca to use a clean configuration by instructing it to use an alternate directory for user preferences i.e. run something like `orca -r -u .`?

**Comment by jpnurmi** [comment]
I installed Ubuntu 20.04 with Orca 3.36.2 and enabled the "Speak object under mouse" setting.

[flutter-orca-ubuntu-20.04.webm](https://user-images.githubusercontent.com/140617/180378990-a369abee-d32b-4ce7-adb5-854dc6d43a7e.webm)

P.S. Even though Orca seems to speak the objects under the mouse fine, I can see some critical warnings in output. I can take a look at this.
```
(counter:4758): Atk-CRITICAL **: 08:45:55.738: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
```

**Comment by jpnurmi** [comment]
@maheshmnj I'm not familiar with Orca internals but perhaps the traceback you posted is unrelated to Flutter because it mentions `Chromium/speech_generator.py` and `web/speech_generator.py`?

**Comment by jpnurmi** [comment]
> P.S. Even though Orca seems to speak the objects under the mouse fine, I can see some critical warnings in output. I can take a look at this.
> 
> ```
> (counter:4758): Atk-CRITICAL **: 08:45:55.738: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
> ```

It seems to be caused by the old AtkText stub in FlAccessibleNode: https://github.com/flutter/engine/pull/34839

**Comment by maheshmnj** [comment]
> I installed Ubuntu 20.04 with Orca 3.36.2 and enabled the "Speak object under mouse" 

Thanks, @jpnurmi enabling "speak object under the mouse". I can confirm the pointer reads flutter widgets under the mouse. And yes the previous logs seem related to other apps and not flutter.


<details>
<summary>logs</summary>

```
Running "flutter pub get" in counter...
Launching lib/main.dart on Linux in debug mode...
** (counter:8644): CRITICAL **: 17:58:00.510: Failed to read XDG desktop portal settings: GDBus.Error:org.freedesktop.portal.Error.NotFound: Requested setting not found
** (counter:8644): CRITICAL **: 17:58:00.513: Failed to read XDG desktop portal settings: GDBus.Error:org.freedesktop.portal.Error.NotFound: Requested setting not found
Connecting to VM Service at ws://127.0.0.1:38009/ItlvOSI0heE=/ws
(counter:8644): Atk-CRITICAL **: 17:58:07.911: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:07.934: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:08.078: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:08.268: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:08.450: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:08.536: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:08.781: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:08.799: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:08.842: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:08.859: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:09.041: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:09.165: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:09.292: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:09.409: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:09.488: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:09.589: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:09.609: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:09.739: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:10.503: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:11.949: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:12.074: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:12.503: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:13.641: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:13.786: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed
(counter:8644): Atk-CRITICAL **: 17:58:13.904: atk_text_get_range_extents: assertion 'start_offset >= 0 && start_offset < end_offset' failed

```
</details>

**Comment by cbracken** [comment]
Fixed by flutter/engine#34839.
